include/libp2p/common/metrics/instance_count.hpp

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef LIBP2P_COMMON_METRICS_INSTANCE_COUNT_HPP
+#define LIBP2P_COMMON_METRICS_INSTANCE_COUNT_HPP
+
+#include <atomic>
+#include <mutex>
+#include <string_view>
+#include <unordered_map>
+
+#define _LIBP2P_METRICS_INSTANCE_COUNT(type, key)                    \
+  struct Libp2pMetricsInstanceCount {                                \
+    static auto &count() {                                           \
+      static auto &count = ([]() -> auto & {                         \
+        auto &state{::libp2p::metrics::instance::State::get()};      \
+        std::lock_guard lock{state.mutex};                           \
```

Comment:
`count()`'s `count` is static local variable, so this lambda with mutex locking is called only once, to insert key-counter pair into map.
Generally LIBP2P_METRICS_INSTANCES_COUNT constructor/destructor is just checking "static guard variable" (was static variable initialized) followed by atomic increment/decrement.

LIBP2P_METRICS_INSTANCES_LIST is heavier and always locks mutex, because it inserts/erases list. Developer chooses whether to keep list of all `type` instances or not (or just count with LIBP2P_METRICS_INSTANCES_COUNT).

Reply:
disabled by default via cmake option